File: C/lesson01-3.c
#include <stdio.h> // printf()
#include <stdlib.h> // abs()
#include <stdint.h>
#include <stdbool.h>
#include "lib/tga.h"

#define BLACK (0xFF000000)
#define WHITE (0xFFFFFFFF)
#define RED   (0xFFFF0000)
#define GREEN (0xFF00FF00)
#define BLUE  (0xFF0000FF)

#define WIDTH 100
#define HEIGHT 100

#define SWAP16(a,b) { uint16_t t = a; a = b; b = t; }

void line(uint32_t *pixels, uint16_t x0, uint16_t y0, uint16_t x1, uint16_t y1, uint32_t color)
{
	bool steep = false;
	if (abs(x0 - x1) < abs(y0 - y1)) {
		SWAP16(x0, y0);
		SWAP16(x1, y1);
		steep = true;
	}
	if (x1 < x0) {
		SWAP16(x0, x1);
		SWAP16(y0, y1);
	}
	for (int x = x0; x < x1; x++) {
		float t = (x - x0) / (float) (x1 - x0);
		int y = y0 * (1. - t) + y1 * t;
		if (steep) {
			pixels[x * WIDTH + y] = color;
		} else {
			pixels[y * WIDTH + x] = color;
		}
	}
}

int main(void)
{
	uint32_t data[WIDTH * HEIGHT];
	for (size_t i = 0; i < WIDTH * HEIGHT; i++) {
		data[i] = BLACK;
	}
	
	line(data, 13, 20, 80, 40, WHITE);
	line(data, 20, 13, 40, 80, RED);
	line(data, 80, 40, 13, 20, BLUE);
	
	tga_save("lesson01-3.tga", data, WIDTH, HEIGHT);
	printf("DONE\n");
	return 0;
}

File: Python/lesson01-3.py
#!/usr/bin/python3

import tga

black = (0, 0, 0)
white = (255, 255, 255)
red = (255, 0, 0)
green = (0, 255, 0)
blue = (0, 0, 255)

def line(buffer, x0, y0, x1, y1, color):
	steep = False
	if abs(x0 - x1) < abs(y0 - y1):
		x0, y0 = y0, x0
		x1, y1 = y1, x1
		steep = True
	if x0 > x1:
		x0, x1 = x1, x0
		y0, y1 = y1, y0
	for x in range(x0, x1):
		t = (x - x0) / (x1 - x0)
		y = y0 * (1 - t) + y1 * t
		if steep:
			buffer[int(x)][int(y)] = color
		else:
			buffer[int(y)][int(x)] = color

# new 100x100 black image
data = [
	[black for x in range(100)] for y in range(100)
]

line(data, 13, 20, 80, 40, white)
line(data, 20, 13, 40, 80, red)
line(data, 80, 0, 13, 20, blue)

data = data[::-1] # flip vertically
# save as TGA
image = tga.Image(data)
image.save("lesson01-3.tga")

print("DONE")